### PR TITLE
feat(iroh-net): cache for crypto keys

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

An attempt to get down the size of a key to something more manageable, see https://github.com/n0-computer/iroh/issues/1673

## Notes & open questions

Size of the cache and TTL. I think 1 minute is fine, if you don't perform any crypto ops on a key for 1 minute it means you are not connected to it. Size is more difficult, but the items are not that big so there is not much harm in having a big size.

Note two: this should help a lot for the derper.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: iroh-net/src/key.rs [hunk 1]
```diff
@@ -38,7 +58,7 @@ impl Serialize for PublicKey {
     where
         S: serde::Serializer,
     {
-        self.public.serialize(serializer)
+        self.public().serialize(serializer)
```
Comment (rklaehn) [inline iroh-net/src/key.rs:61]: Why can't this be self.0.serialize?

File: iroh-net/Cargo.toml [hunk 2]
```diff
@@ -78,6 +78,8 @@ tracing-subscriber = { version = "0.3", features = ["env-filter"], optional = tr
 # metrics
 iroh-metrics = { version = "0.7.0", path = "../iroh-metrics", default-features = false }
 bounded_join_set = "0.1.0"
+lazy_static = "1.4.0"
```
Comment (dignifiedquire) [inline iroh-net/Cargo.toml:81]: `once_cell` please, we already use that

File: iroh-net/src/key.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-net/src/key.rs:61]: you need to use `serialize_bytes`, see https://github.com/dalek-cryptography/ed25519-dalek/blob/main/src/verifying.rs#L630

File: iroh-net/Cargo.toml [hunk 3]
```diff
@@ -79,6 +79,8 @@ tracing-subscriber = { version = "0.3", features = ["env-filter"], optional = tr
 # metrics
 iroh-metrics = { version = "0.7.0", path = "../iroh-metrics", default-features = false }
 bounded_join_set = "0.1.0"
+ttl_cache = "0.5.1"
+serde_bytes = "0.11.12"
```
Comment (rklaehn) [inline iroh-net/Cargo.toml:83]: we can also write the weird visitors ourselves if you don't want the dependency. But it's small and from @dtolnay, so I'd rather not.

**Review by flub — COMMENTED** [review]
Have you considered putting the cache into the magicsock struct rather than having it global?  It would make the change more invasive probably, but you have more flexibility.  E.g. you could potentially allow customising it etc.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**